hubble: Use a single string to configure the server address

michi-covalent opened the pull request:
- Rename the option to `--hubble-listen-address` to make it clear that
  Hubble listens to at most one additional address.
- Remove an info message that doesn't provide any addtional info:
    ```
      level=info msg="Starting local Hubble server" address="unix:///var/run/cilium/hubble.sock" subsys=hubble
      level=info msg="Starting Hubble server" address=":4244" subsys=hubble
    - level=info msg="Starting gRPC server on listener" listener="unix:///var/run/cilium/hubble.sock" subsys=hubble
    - level=info msg="Starting gRPC server on listener" listener=":4244" subsys=hubble
    ```
- Use port 4244 in the documentation since this is the port that Hubble
  relay expects.

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

michi-covalent committed:
hubble: Use a single string to configure the server address

- Rename the option to `--hubble-listen-address` to make it clear that
  Hubble listens to at most one additional address.
- Remove an info message that doesn't provide any addtional info:
    ```
      level=info msg="Starting local Hubble server" address="unix:///var/run/cilium/hubble.sock" subsys=hubble
      level=info msg="Starting Hubble server" address=":4244" subsys=hubble
    - level=info msg="Starting gRPC server on listener" listener="unix:///var/run/cilium/hubble.sock" subsys=hubble
    - level=info msg="Starting gRPC server on listener" listener=":4244" subsys=hubble
    ```
- Use port 4244 in the documentation since this is the port that Hubble
  relay expects.

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

michi-covalent committed:
hubble: Point hubble-ui to hubble-relay service

- Configure hubble-ui to connect to hubble-relay:80. Now hubble-relay is
  responsible for retrieving flows from Cilium instances.
- Remove hubble-grpc service.

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

soumynathan reviewed (commented):
LGTM

soumynathan reviewed (commented):
LGTM

rolinh commented on daemon/cmd/hubble.go:
```diff
@@ -131,17 +124,22 @@ func (d *Daemon) launchHubble() {
+		}
 		srv, err := server.NewServer(logger,
-			serveroption.WithListeners(addresses),
+			serveroption.WithListeners([]string{address}),
```
Thinking about it, I don't see a use-case for the ability to listen on an additional Unix domain socket. Thus, I propose the following: remove `WithListeners()` in `pkg/hubble/server/serveroption/option.go` and change this call for `WithTCPListener()` (it will fail if it's not a TCP socket).

I think we should also get rid of the `Listeners` map in `Options` and simply have `TCPListener` and `UnixSocketListener` members  as keeping a map for at most 2 values is overkill.

michi-covalent replied:
will address in a follow-up pr.

michi-covalent replied:
or in this pr: https://github.com/cilium/cilium/pull/11330/commits/9deccf928900ba79612918570665687ab4c950a8

rolinh reviewed (changes requested):
It looks good but I think we can simplify this further.

tklauser commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -432,16 +432,13 @@ global:
-    #
-    #   --set global.hubble.listenAddresses={:50051,:50052}
-    listenAddresses: []
+    # Set this field ":4244" if you are enabling hubble-relay, as it assumes htat Hubble is listening
```
Nit: small typo
```suggestion
    # Set this field ":4244" if you are enabling hubble-relay, as it assumes that Hubble is listening
```

michi-covalent replied:
fixed

michi-covalent committed:
hubble: Remove serveroption.WithListeners

Now that Hubble has 1 unix domain socket and at most 1 TCP socket to
serve, we can simply use `With{UnixSocket,TCP}Listener` options.

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

rolinh commented on daemon/cmd/hubble.go:
```diff
@@ -131,17 +124,22 @@ func (d *Daemon) launchHubble() {
+	address := option.Config.HubbleListenAddress
+	if address != "" {
+		// TODO: remove warning once mutual TLS has been implemented
+		if !strings.HasPrefix(address, "unix://") {
```
This `if` check can now be removed as a unix domain socket is no longer expected as an additional listen address (but leave the warning).

michi-covalent replied:
fixed

rolinh commented on pkg/hubble/server/server.go:
```diff
@@ -68,13 +65,23 @@ func (s *Server) initGRPCServer() {
+			if err := s.srv.Serve(s.opts.TCPListener); err != nil {
+				s.log.WithError(err).
+					WithField("address", s.opts.TCPListener.Addr().String()).
+					Error("failed to start grpc server")
```
nit: to avoid code duplication, you could keep the loop.
```go
	for _, listener := range []net.Listener{s.opts.UnixSocketListener, s.opts.TCPListener} {
		if listener != nil {
			go func(listener net.Listener) {
				if err := s.srv.Serve(listener); err != nil {
					s.log.WithError(err).
						WithField("address", listener.Addr().String()).
						Error("failed to start grpc server")
				}
			}(listener)
		}
	}
```

michi-covalent replied:
fixed

rolinh reviewed (commented):
Awesome! Please, just remove the `if` check as it is superfluous now.

rolinh reviewed (approved):
LGTM, thanks!

rolinh commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -91,19 +68,21 @@ func WithUnixSocketListener(path string) Option {
 		unix.Unlink(socketPath)
 		socket, err := net.Listen("unix", socketPath)
 		if err != nil {
+			fmt.Println("MICHI 1", err)
```
:thinking:

michi-covalent replied:
ok that's embarrassing :)

rolinh commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -91,19 +68,21 @@ func WithUnixSocketListener(path string) Option {
 		}
 		if os.Getuid() == 0 {
 			if err := api.SetDefaultPermissions(socketPath); err != nil {
+				fmt.Println("MICHI 2", err)
```
Where you debugging something in particular?

rolinh reviewed (changes requested):
Not sure if I missed the leftover print debugging statement during my last review or if you added them with a force push but I'm pretty sure we don't want to keep them :)
